src/fairseq2/models/llama/integ.py

```diff
@@ -38,3 +53,55 @@ def convert_to_reference_checkpoint(checkpoint: dict[str, Any]) -> dict[str, Any
     }
 
     return convert_model_state_dict(state_dict, key_map)
+
+
+def convert_to_huggingface_config(arch: str, config: LLaMAConfig) -> dict[str, Any]:
+    """Convert Llama's config to a dict mirroring Huggingface's format"""
+
+    def compute_intermediate_size(
+        n: int, ffn_dim_multiplier: float = 1, multiple_of: int = 256
+    ) -> int:
+        """From: https://github.com/huggingface/transformers/blob/82fcac0a7e40dc6cc5e3121d714b9b16775293ad/src/transformers/models/llama/convert_llama_weights_to_hf.py#L171"""
+        return multiple_of * (
+            (int(ffn_dim_multiplier * int(8 * n / 3)) + multiple_of - 1) // multiple_of
+        )
+
+    def is_llama_3_2(arch: str) -> bool:
+        # TODO: this seems too britle
+        return "llama3_2_" in arch
```

Comment:
Yeah, agree that this looks a bit hacky, but I think it is fine for now.

Reply:
To preface this, this PR yields exact parity with the current conversion process from our checkpoints to vLLM-compatible HF checkpoints (so: same config.json produced, and I've checked for generation equality on a dataset of prompts with vLLM).


But I think me needing the hack made me find a small inconsistency present on main right now:

RoPE scaling params in the conversion script depend on the Llama version (32.0 factor if Llama version 3.2 for example) but fairseq2 hardcodes these differently (factor is always 8.0):

https://github.com/facebookresearch/fairseq2/blob/03511c2d08fcae5946638a900138d3bc5070370b/src/fairseq2/models/llama/factory.py#L283C9-L286C26

So if we follow the intended workflow (either the current double conversion or simply writing the config.json), we end up with different RoPE parameters in vLLM inference than in training.

Reply:
Interesting. When I check the reference implementation, I see that the scale factor is constant for all 3.x (https://github.com/meta-llama/llama-models/blob/main/models/llama3/reference_impl/model.py#L47). Do you know where that scale factor originates from?